core/src/main/java/org/infinispan/notifications/cachelistener/cluster/ClusterListenerReplicateCallable.java

```diff
@@ -96,11 +97,30 @@ public Void call() throws Exception {
                }
                if (!alreadyInstalled) {
                   RemoteClusterListener listener = new RemoteClusterListener(identifier, origin, distExecutor, cacheNotifier,
-                                                                           cacheManagerNotifier, eventManager, sync);
+                                                                             cacheManagerNotifier, eventManager, sync);
                   // We only need to use the post cache event filter when a filter is provided, since the remote cluster listener
                   // will ignore pre events anyways.  This is mostly so a user filter is not notified of the pre event.
-                  cacheNotifier.addListener(listener,
-                        filter != null ? new CompositeCacheEventFilter(new PostCacheEventFilter(), filter) : null, converter);
+                  CacheEventFilter<K, V> actualFilter = null;
+                  if (filter instanceof CacheEventFilterConverter) {
+                     final CacheEventFilterConverter<K, V, ?> filterConverter = (CacheEventFilterConverter<K, V, ?>) filter;
+                     actualFilter = new AbstractCacheEventFilterConverter<K,V, Object>() {
```

Comment:
It would not do that until it's needed.  A CompositeCacheEventFilterConverter would be a strange beast. It should take the output of first filter and feed it to the next (unless returned value was null). So that would mean C==V for CompositeCacheEventFilterConverter and all its inner filters :)